---
title: 'Open Virome: Earths viruses in your palm'
output:
  html_document:
    df_print: paged
  html_notebook:
    code_folding: hide
params:
  palmid.version: '0.0.7'
  search_type: "SEARCH"            # "LIST" or "STAT" or "SEARCH"
  virome_search_term: '%Eimeria%'  # Scientific_name search term
  virome_deplete_term: 'Candidatus' # Deplete this term from matches, or '' for none
  input.path: inst/extdata/candida # <prefix>.stat OR <prefix>.list
  control_type: 'BIOPROJECT'       # '' (N/A), 'SEARCH', 'LIST', or 'BIOPROJECT'
  control_search_term: ''          # Set '' to exclude control search
  control.path: inst/extdata/yeast
  export.cytoscape: TRUE
  prod.run: no
editor_options:
  markdown:
  wrap: 72
---
# Open Virome Markdown  {.tabset}

```{r}
# Open Virome TODO:
# - [ ] Report virus as the "rates of occurrence" within search queries
# - [ ] Include metrics for Alpha and Beta-Diversity (conceptual development needed)
# - [ ] Incorporate Page_Rank for sOTU and Runs (seperately) to rank output tables
# - [ ] Report RdRp-network and explore network enrichment for this
```

### Analysis Overview {.tabset}
 
```{r setup, include=FALSE}
# RMarkdown Setting Initialization
# Command Line Interface ----
#knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{css, echo=FALSE}
/* Move code folding buttons to the left */
div.col-md-12 .pull-right {
  float: left !important
}
```

```{r production, include=FALSE, warning=FALSE}
# Run as Production Pipeline or Development
production.version = params$prod.run

if (production.version) {
  # Load stable palmid package
  library('palmid', quietly = T)
  data("palmdb")

  # Disable warnings
   defaultW <- getOption("warn")
   options(warn = -1)
} else {
  # Warnings variable
  defaultW <- getOption("warn")
  options(warn = -1)

  # Compile a new palmid from source
  roxygen2::roxygenise()
  load("data/palmdb.RData")
}

```

```{r initaialize, include = FALSE}
# INITIALIZE OPEN VIROME WORKSPACE -------------------

# Establish Serratus server connection
con <- SerratusConnect()

# Input files
# Generated within `palmid` container
# input.path parameter is defined in YAML header
# which is exposed to CLI
input.path   <- params$input.path
control.path <- params$input.path

if (is.null(input.path)) {
  stop("Error: No input provided.")
}

  input.list      <- paste0(input.path, '.list')    # Virome search list file
  control.list    <- paste0(control.path, '.list')  # Control search list file
  input.virome    <- paste0(input.path, '.virome')  # preamde palm_virome query csv
  input.stat      <- paste0(input.path, '.stat' )   # STAT-based query

# Output HTML Report
output.html   <- paste0(input.path, '.html')

# Parameters
# Search Term of SRA "Scientific Name" field,
# uses SQL "like" operator, which allows for regex
# expression i.e. 'Saccharomyces%'
virome_search_term  <- params$virome_search_term
virome_deplete_term <- params$virome_deplete_term

# Control Set
control_type  <- params$control_type
control_search_term <- params$control_search_term

# Exclude control logical variable
if (control_type == '' ){
  doControl = FALSE
} else {
  doControl = TRUE
}

# Include Cytoscape network plot export
export.cytoscape <- params$export.cytoscape

# LOOK AND FEEL -----------------------

# Control / Test set colors
ui.setcol <- c('gray50', 'cornflowerblue')
```

```{r initialize message}
# Print out version
cat( paste0("palmID Version: ", params$palmid.version) )
cat( paste0("-- Search Type: ", params$search_type) )
cat( paste0("-- Taxa search term: ", virome_search_term))
if (doControl){   cat( paste0("-- Control search term: ", control_search_term )) }

```

```{r, echo = FALSE}
# Import virome data

if (params$search_type == "LIST"){
  # Retrieve Virome using a SRA Run vector
  virome.runs  <- read.csv(input.list)[,1]
  #virome.runs <- XXXXXX # Manual Override
  virome.df    <- get.palmVirome(run.vec = virome.runs)
} else if (params$search_type == "SEARCH"){
  # Retrieve virome using search term
  virome.df    <- get.palmVirome(org.search = virome_search_term)
  
  # Deplete terms (if available)
  deplete.runs <- grep( virome_deplete_term, virome.df$scientific_name, ignore.case = T)
  if (length(deplete.runs > 0)){
    virome.df    <- virome.df[ -deplete.runs, ]
  } else {
    # No runs to deplete
  }
  
  virome.runs  <- virome.df$run
  
} else if (params$search_type == "STAT"){
  # Retrieve Virome using a SRA-STAT table
  virome.stat  <- read.csv(input.stat, header = T)
  virome.stat  <- virome.stat[ virome.stat$total_count > 10000, ]
  virome.runs  <- virome.stat$acc
  
  virome.df    <- get.palmVirome(run.vec = virome.runs)
  virome.df    <- merge(virome.df, virome.stat, by.x = 'run', by.y = 'acc')
} else {
  stop('Unknown Open Virome Run mode. Use: "LIST" or "SEARCH" or "STAT"')
}

# Melt virome, grouping by "sOTU"
  virx.df   <- melt.virome( virome.df )
  
# Import "Control" Virome
if (control_type == "LIST"){
  # Retrieve Virome using a SRA Run vector
  negVirome.df <- get.negativeVirome(run.vec = virome.runs )
} else if (control_type == "SEARCH"){
  # Retrieve virome using search term
  negVirome.df <- get.negativeVirome(org.search = virome_search_term)
  
  #} else if (control_type == "STAT"){
  # Retrieve Virome using a SRA-STAT table (Not implemented)

} else if (control_type == "BIOPROJECT"){
  neg.virome.runs <- get.sraProj( run_ids <- virome.df$run,
                                  exclude.input.runs = TRUE,
                                  con = con)
  # If no negative virome samples found, negVirome is NA
  if (length(neg.virome.runs$run_id) == 0){
    negVirome.df <- NA
    doControl <- FALSE
  } else {
    negVirome.df <- get.negativeVirome(run.vec = neg.virome.runs$run_id)
    # Melt virome, grouping by "sOTU"  
    negv.df   <- melt.virome( negVirome.df )  
  }
} else {
  stop('Unknown Open Virome Run mode. Use: "LIST" or "SEARCH" or "BIOPROJECT"')
}
  

# # Import Control data
# if (doControl){  
#   if (params$search_type == "LIST"){
#     # Retrieve Control using a SRA Run vector
#     control.runs  <- read.csv(input.list)[,1]
#     control.df    <- get.palmVirome(run.vec = control.runs)
#     negControl.df <- get.negativeVirome(run.vec = control.runs )
#   } else {
#     # Retrieve control using search term
#     control.df    <- get.palmVirome(org.search = control_search_term)
#     negControl.df <- get.negativeVirome(org.search = control_search_term)
#   }
#   # Melt virome, grouping by "sOTU"
#    ctrl.df   <- melt.virome( control.df )
# }

# View(virx.df[ (virx.df$n > 5), ])
# View(virome.df[ virome.df$sotu == 'u196227', ])  
  
```

### Download Virome Data Tables {#download-data}

```{r, echo = FALSE}
options(warn = -1)

virome.df  %>%
  downloadthis::download_this(
    output_name = paste0(input.path, '_virome_full.csv'),
    output_extension = ".csv",
    button_label = "Download virome_full.csv",
    button_type = "success",
    has_icon = TRUE,
    icon = "fa fa-file-export"
  )

virx.df  %>%
  downloadthis::download_this(
    output_name = paste0(input.path, '_virome_summary.csv'),
    output_extension = ".csv",
    button_label = "Download virome_sumnmary.csv",
    button_type = "success",
    has_icon = TRUE,
    icon = "fa fa-file-export"
  )

```

### Virome Run Summary {#run-module}

```{r, echo = FALSE}
library(ggplot2)

# Initialize vorgx.df
# Scientific Name - Virus Taxonomy count frame
# with ordered factors for plotting

makeTop10 <- function( invec , top.n = 10){
  # For an input character vector invec
  # returns invec as a  factor
  # of only top 10 entries
  # Remaining entries named "Other"
  # levels are rank-ordered for plotting
  #
  # Convert to character
  invec <- as.character(invec)
  # Count entries, return top-10
  t10.entries <- table( invec )
  t10.entries <- t10.entries[ rev( order(t10.entries) ) ]
  t10.entries <- rownames(t10.entries)[1:top.n]
  
  # Refactor input vector to only top-10 entries
  # and levels ordered by top-10 order
  invec2 <- invec
  invec2[ !(invec2 %in% t10.entries) ] <- "Other"
  invec2 <- factor(invec2,
                   levels = c(t10.entries, "Other"))
  
  return(invec2)

}
# # Select top 10 terms in Scientific Name and Virus Family
# # for plot, create ordered vector by count
# t10.org <- virome.df %>%
#   count(scientific_name, sort = TRUE) %>%
#   select(scientific_name) %>% slice_head(n = 10)
#   t10.org <- as.character(t10.org$scientific_name)
# 
# t10.fam <- virome.df %>%
#   count(tax_family, sort = TRUE) %>%
#   select(tax_family) %>% slice_head(n = 10)
# t10.fam <- as.character(t10.fam$tax_family)

# Summarize (count) for plotting
# organism and virus family
```

```{r, echo = FALSE}
# Create modified Virome data.frame

if (doControl){
  virome.df2 <- bind_rows( virome.df, negVirome.df)
} else {
  virome.df2 <- virome.df
}
  # scientific name, simplify to top 10
  virome.df2$scientific_name <- makeTop10(virome.df2$scientific_name)
  # virus family taxonomy, simplify to top 10
  virome.df2$tax_family      <- makeTop10(virome.df2$tax_family)
  
# Create virome organism summary data.frame
vorgx.df <- virome.df2 %>% dplyr::count(scientific_name, tax_family, node_qc, sort = TRUE)
  vorgx.df$node_qc[ is.na(vorgx.df$node_qc) ] <- FALSE
  vorgx.df$scientific_name <- factor(vorgx.df$scientific_name,
                                     levels = rev( levels(virome.df2$scientific_name)))
  colnames(vorgx.df) <- c('scientific_name', 'tax_family', 'vRNA', 'n')
  vorgx.df$tax_family      <- factor(vorgx.df$tax_family,
                                     levels = levels(virome.df2$tax_family ))

# Virome bar organism bar plot
virome.org <- ggplot(vorgx.df, aes( scientific_name, n,
                                    fill = vRNA)) +
  geom_bar( stat = 'identity') +
  coord_flip() +  theme_bw() +
  xlab("Scientific Name (metadata)") +
  ylab("Count SRA Runs") + facet_wrap(~vRNA) +
  scale_fill_manual( values = ui.setcol)

```

```{r, fig.height = 4.5, fig.width = 9}
# Virome bar plot of "Scientific Name" Fields
plotly::hide_colorbar( virome.org )
```


```{r, echo = FALSE}
# Retrieve SRA Meta-data for virome data.frame
sra.df <- get.sraMeta( virome.df2$run, con = con, ordinal = TRUE)

# SRA Meta-data aggregation
sra.df$vRNA <- "Control"
sra.df$vRNA[ (sra.df$run_id %in% virome.df$run) ] <- "Target"

# # Run Data-type -- R VERSION
# sra.data <- sra.df %>%
#             count(vRNA, library_strategy)
# sra.data.nt <- aggregate(bases ~ vRNA + library_strategy, sra.df, sum)
# 
# sra.data    <- sra.data[ order(sra.data$vRNA,       sra.data$library_strategy), ]
# sra.data.nt <- sra.data.nt[ order(sra.data.nt$vRNA, sra.data.nt$library_strategy), ]
# 
# sra.data$Gbp <- (sra.data.nt$bases) / 1e9
# 
# sra.dataType <- ggplot(sra.data, aes( x = library_strategy, n,
#                                       fill = Gbp)) +
#   geom_bar( stat = 'identity') +
#   scale_fill_viridis_b() +
#   geom_col() +
#   geom_text(aes(label = paste0(n, " runs")), vjust = 1.5, colour = "black") +
#   scale_y_log10() + coord_polar("x", start = 0) +
#   theme_bw() + ggtitle("SRA Dataset Types") +
#   theme(aspect.ratio = 1 ) + 
#   facet_wrap(~vRNA, ncol = 2)
# 
# sra.dataType2 <- ggplot(sra.data, aes( x = library_strategy, Gbp,
#                                       fill = n )) +
#   geom_bar( stat = 'identity') +
#   scale_fill_viridis_b() +
#   geom_col() +
#   geom_text(aes(label = paste0(round(Gbp, 1), " Gbp")), vjust = 1.5, colour = "black") +
#   scale_y_log10() + coord_polar("x", start = 0) +
#   theme_bw() + ggtitle("SRA Dataset Types") +
#   theme(aspect.ratio = 1 ) + 
#   facet_wrap(~vRNA, ncol = 2)


# Run Data-type - OV VERSION
sra.data <- sra.df %>%
            count(vRNA, library_strategy)
sra.data.nt <- aggregate(bases ~ vRNA + library_strategy, sra.df, sum)

sra.data    <- sra.data[ order(sra.data$vRNA,       sra.data$library_strategy), ]
sra.data.nt <- sra.data.nt[ order(sra.data.nt$vRNA, sra.data.nt$library_strategy), ]

sra.data$Gbp <- (sra.data.nt$bases) / 1e9

sra.dataType <- ggplot(sra.data, aes( x = library_strategy, n,
                                      fill = vRNA)) +
  geom_bar( stat = 'identity') +
  scale_fill_manual( values = ui.setcol ) +
  geom_col() +
  geom_text(aes(label = paste0(n, " runs")), vjust = 1.5, colour = "black") +
  scale_y_log10() + coord_polar("x", start = 0) +
  theme_bw() + ggtitle("Dataset Types -- Count") +
  theme(aspect.ratio = 1, legend.position = "none") + 
  facet_wrap(~vRNA, ncol = 2)

sra.dataType2 <- ggplot(sra.data, aes( x = library_strategy, Gbp,
                                      fill = vRNA )) +
  geom_bar( stat = 'identity') +
  scale_fill_manual( values = ui.setcol ) +
  geom_col() +
  geom_text(aes(label = paste0(round(Gbp, 1), " Gbp")), vjust = 1.5, colour = "black") +
  scale_y_log10() + coord_polar("x", start = 0) +
  theme_bw() + ggtitle("Dataset Types -- Gbp") +
  theme(aspect.ratio = 1, legend.position = "none") + 
  facet_wrap(~vRNA, ncol = 2)

```

```{r,  fig.height = 8, fig.width = 10, retina = 3}
# SRA Datasets Type
sra.dataType
sra.dataType2
```


```{r, echo = FALSE}
# BioProject Sizes
bp.target.n <- sra.df %>%
               count(vRNA, bioproject, sort = TRUE) %>%
               filter( vRNA == 'Target' )

bp.total.n <- sra.df %>%
               count(bioproject, sort = TRUE)

bp.df      <- merge(bp.total.n, bp.target.n[ , c('bioproject', 'n')], all.x = TRUE, by = 'bioproject')
  colnames(bp.df) <- c('bioproject', 'bp_n', 'target_n')
  bp.df$target_n[ is.na(bp.df$target_n) ] <- 0
  bp.df$perc_target <- 100* bp.df$target_n / bp.df$bp_n

bp.Size <- ggplot(bp.df, aes( x = bp_n, perc_target, bp = bioproject, color = perc_target)) +
  geom_jitter(width = 0, height = 2,
              alpha = 0.5) +
  theme_bw() + scale_x_log10() +
   scale_colour_continuous(type = "gradient") +
  ggtitle("Runs per BioProject vs. Percent BioProject-Runs in Target Set (%)") +
  xlab('BioProject Size (n runs)') + ylab('Percent BioProject in Test Set') +
  theme(aspect.ratio = 1, legend.position = "none")

bp.hist <- ggplot(bp.df, aes( x = bp_n )) +
  geom_histogram( n = 60 ) +
  theme_bw() + scale_x_log10() +
  ggtitle("BioProject size") +
  xlab('(n runs)') + ylab('count') +
  theme(aspect.ratio = 0.6, legend.position = "none")


bp.hist2 <- ggplot(bp.df, aes(perc_target)) +
  geom_histogram( n = 25 , fill = hcl.colors(30, "Blues2", rev = T)) +
  theme_bw() +
  ggtitle("% of BioProject in Target") +
  xlab('Target (%)') + ylab('count') +
  theme(aspect.ratio = 0.6, legend.position = "none")
bp.hist

```

```{r,  fig.height = 8, fig.width = 10, retina = 3}
# SRA Datasets Type
plotly::hide_legend( bp.Size )
bp.hist
bp.hist2
```

### Virus Family Summary {#family-module}

```{r, echo = FALSE}
# Create virome family - Run summary (top 20)
virFam.nrun <- virome.df[,  c('tax_family', 'run')]
virFam.nrun$tax_family <- makeTop10(virFam.nrun$tax_family, top.n = 20)
virFam.nrun <- unique(virFam.nrun[,  c('tax_family', 'run')]) %>% count(tax_family)
virFam.nrun$set <- 'n_runs'

# Create virome family - sOTU summary (top 20)
virFam.sotu <- virome.df[,  c('tax_family', 'sotu')]
virFam.sotu$tax_family <- makeTop10(virFam.sotu$tax_family, top.n = 20)
virFam.sotu <- unique(virFam.sotu[,  c('tax_family', 'sotu')]) %>% count(tax_family)
virFam.sotu$set <- 'n_sotu'

virFam.df <- rbind(virFam.nrun, virFam.sotu)

# Virome Family count bar plot
virFam.n <- ggplot(virFam.df, aes( tax_family, n)) +
  geom_bar( stat = 'identity') +
  coord_flip() +  theme_bw() +
  xlab("Taxonomic Family") +
  ylab("Count") + facet_wrap(~set)
  scale_fill_manual( values = c('gray50', 'cornflowerblue'))

virFam.n
```
```{r, echo = FALSE}
# Scatter plot of Virus Family in `n_runs` vs. `n_sotu`
virFam.nrun <- unique(virome.df[,  c('tax_family', 'run')]) %>% count(tax_family)
  colnames(virFam.nrun) <- c("tax_family", "n_run")
virFam.sotu <- unique(virome.df[,  c('tax_family', 'sotu')]) %>% count(tax_family)
  colnames(virFam.sotu) <- c("tax_family", "n_sotu")

virFam.df2 <- merge( virFam.nrun, virFam.sotu, by = "tax_family")

# Virome Family count Scatter plot
virFam.n <- ggplot(virFam.df2, aes( n_run, n_sotu, color = tax_family)) +
  geom_point() +
  theme_bw() +
  theme(aspect.ratio = 1, legend.position = "none") +
  xlab("Count (Runs)") + ylab("Count (sOTU)")

```

```{r,  fig.height = 8, fig.width = 10, retina = 3}
# SRA Datasets Type
plotly::hide_legend( virFam.n )
```


```{r, echo = FALSE}
# Create a data.frame to plot a heatmap of Virus Family versus BioProject
# where the heatmap value is the Percent of the BioProject positive for that Virus Family

virFam.df2<- virome.df[,  c('tax_family', 'bio_project')]
virFam.df2$tax_family <- makeTop10(virFam.df2$tax_family, top.n = 20)
virFam.df2 <- table(virFam.df2)
  bpTop   <- colnames(virFam.df2)
  bpTop.n <- as.numeric(bp.total.n$n[ match(bpTop, bp.total.n$bioproject, nomatch = NA) ])
  bpTop.n <- bpTop.n[!is.na(bpTop.n)]
  
# Divide each matrix-row by vector
# Form: t( t(M) / vec )
virFam.df2 <- t( t(virFam.df2) / bpTop.n)
virFam.df2 <- round( 100 * virFam.df2, 2 )
virFam.df2 <- as.matrix(virFam.df2)
  
# Virome Family count bar plot
virFam.bp <- gplots::heatmap.2(virFam.df2, trace = "none", 
                               breaks = c(0,1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
                               density.info="none",
                               col = c("black", viridis(10, option = "A")),
                               key.title = "",
                               key.xlab = "Percent BioProject in Target",
                               margins = c(10,10), sepcolor=NULL)

```

### Network Analysis {#network-module}

```{r, echo = FALSE}
library(vegan)
library(viridis)
library(igraph)
library(RCy3)
library(reshape2)

# Virome diversity per species
# measurement of alpha and beta diversity of two things
# diversity of 
# vcounts.df <- table( virome.df$scientific_name, virome.df$sotu )

# Create virome family heatmap (top 20 Taxa)
#vir.graph <- graph.virome(virome.df, nodes = "run", edges = "sotu",
#                          node.col = "bio_project", edge.col = "sotu")


# Create a bipartite (sOTU + Run) graph based on the virome
vir.graph2 <- graph.virome2(virome.df)

```

```{r}
# Plot Network (add conditional to avoid plotting large networks)
# Interactive plot with pull-down of components of the network

if ( length( V(vir.graph2) ) < 1000 & length( E(vir.graph2) ) < 5000 ) {
  plot.igraph(vir.graph2,
              layout = layout_nicely,
              vertex.size = 5,
              vertex.label = NA,
              vertex.color = V(vir.graph2)$type,
              arrow.mode = "-",
              rescale = TRUE) 
} else {
  print("Skipping network plot, >1000 Nodes or >5000 edges")
}

#tkplot(     vir.graph, layout = layout_with_kk,
#            vertex.size = 10,
#            vertex.color = V(vir.graph)$vcol,
#            edge.width = E(vir.graph)$Weights,
#            arrow.mode = E(vir.graph)$arrow.mode)

if (export.cytoscape == TRUE ){
  # Export to CytoScape (must be open)
  createNetworkFromIgraph(vir.graph2,"virGraph2")
}

```

```{r}
# Component-wise statistics for the network
component.stats <- function(g){
  # For each "component" in a graph, calculates
  # cs : Component stats data.frame
  cs <- data.frame( component = "nil",
                    n_sotu = 0,
                    n_run  = 0,
                    n_edge = 0,
                    D_sotu = 0,
                    D_run  = 0,
                    Vrich  = 0,
                    Dia    = 0)
  
  # Calculate summary data.frame for vir.graph per component
  comp.names <- unique(V(g)$component)
  n.comp <- length( comp.names )
  L.index <- (V(g)$type == TRUE) # TRUE is sOTU nodes
    
  for (i in seq(1, n.comp)){
    # Vertices in subgraph of each component
    V.index <- (V(g)$component == comp.names[i])
      I.sotu <- which(V.index & L.index)
      I.runs <- which(V.index & !L.index)
    
    subg <- induced_subgraph(g, V.index)
      subI.sotu <- (V(subg)$type == TRUE)
      subI.runs <- (V(subg)$type == FALSE)
    
    # cs for ith component
    cs.i <- data.frame( component = comp.names[i],
                    n_sotu = length(I.sotu),
                    n_run  = length(I.runs),
                    n_edge = length(E(subg)$run),
                    D_sotu = mean( degree(subg)[subI.sotu] ),
                    D_run  = mean( degree(subg)[subI.runs] ),
                    Vrich  = sum( V(subg)$vrich[subI.sotu] ),
                    Dia    = diameter( subg ) )
    
    cs <- rbind(cs, cs.i)
    }
  
  cs <- cs[-1,] # Remove null-initialization row
  cs <- cs[ order( as.numeric( as.character( cs$component ) )), ]
  cs$component <- factor(cs$component)
  return(cs)
}


# Calculate stats per component
cs.df <- component.stats(vir.graph2)
  cs.df$n_nodes   <- cs.df$n_sotu + cs.df$n_run
  cs.df$perc_sotu <- 100 *   cs.df$n_sotu / cs.df$n_nodes
```

```{r}
# Plot variables
netlim <- c(1, max(with(cs.df, c(n_nodes, n_edge)))) # Network size limit (log)
nodelim <- c(0, max(with(cs.df, c(n_sotu, n_run)))) # Node Range Limits
dlim    <- range(with(cs.df, c(D_sotu, D_run)))     # Degree Range Limits

# Whole Vertex vs. Edge size
cs.plot0 <- ggplot(cs.df, aes(n_nodes, n_edge,
                             label = component,
                             fill = component)) +
  geom_abline(slope = 1, intercept = 0, color = 'gray50') +
  geom_label(colour = "white", fontface = "bold") +
  theme_bw() + theme(legend.position = "none") +
  scale_x_log10() + scale_y_log10() +
  scale_fill_manual(values = turbo(length(cs.df$component)))  +
  coord_cartesian(xlim = netlim , ylim = netlim ) + 
  xlab('Number of Nodes') + 
  ylab('Number of Edges')
cs.plot0

# Run vs OTU
cs.plot1 <- ggplot(cs.df, aes(n_run, n_sotu,
                             label = component,
                             size = log10(n_edge),
                             fill = component)) +
  geom_abline(slope = 1, intercept = 0, color = 'gray50') +
  geom_label(colour = "white", fontface = "bold") +
  theme_bw() + theme(legend.position = "none") +
  scale_fill_manual(values = turbo(length(cs.df$component))) +
  coord_cartesian(xlim = nodelim, ylim = nodelim) + 
  xlab('Number of Runs (nodes)') + 
  ylab('Number of sOTU (nodes)')
cs.plot1

# Cumulative Virome Enrichment per Component
cs.plot2 <- ggplot(cs.df, aes(n_sotu, Vrich,
                             label = component,
                             size = log10(n_edge),
                             fill = component)) +
  geom_abline(slope = 1, intercept = 0, color = 'gray50') +
  geom_label(colour = "white", fontface = "bold") +
  theme_bw() + theme(legend.position = "none") +
  scale_fill_manual(values = turbo(length(cs.df$component))) +
  coord_cartesian(xlim = nodelim, ylim = nodelim) + 
  xlab('Number of sOTU (nodes)') + 
  ylab('Cumulative Virome Enrichment')
cs.plot2


# Degree of Run vs OTU
cs.plot3 <- ggplot(cs.df, aes(D_sotu, D_run,
                             label = component,
                             size = log10(n_edge),
                             fill = component)) +
  geom_abline(slope = 1, intercept = 0, color = 'gray50') +
  geom_label(colour = "white", fontface = "bold") +
  theme_bw() + theme(legend.position = "none") +
  scale_fill_manual(values = turbo(length(cs.df$component)))  +
  coord_cartesian(xlim = dlim, ylim = dlim) + 
  xlab('Mean Runs per sOTU (Degree)') +
  ylab('Mean sOTU per Run (Degree)') 
cs.plot3


```

```{r}
# Create summary data.
i.sotu <- V(vir.graph2)$type
V.df <- data.frame( sotu   = V(vir.graph2)$name[ i.sotu  ],
                    vrich  = V(vir.graph2)$vrich[ i.sotu ],
                    pr     = V(vir.graph2)$pr[ i.sotu    ],
                    vrank  = V(vir.graph2)$vrank[ i.sotu ],
                    vexact = V(vir.graph2)$v.exact[ i.sotu ],
                    nruns  = degree(vir.graph2)[ i.sotu ] )
                    
V.df <- V.df[ order(V.df$vrank, decreasing = TRUE), ]

V.plot <- ggplot( V.df, aes(pr, vrich,
                            fill = vrank,
                            label = sotu)) +
  geom_label(colour = "white", fontface = "bold") +
  #geom_point() +
  scale_fill_viridis(option = "plasma") +
  theme_bw() +
  xlab('Page Rank (sOTU)') +
  ylab('V-enrichment (sOTU)') 
V.plot
 
# V2.plot <- ggplot( V.df, aes(nruns, vexact,
#                              fill  = vrank,
#                              label = sotu)) +
#   geom_label(colour = "white", fontface = "bold") +
#   geom_abline(color = 'red', slope = 0, intercept = -log10(0.001)) +
#   scale_fill_viridis(option = "plasma") +
#   theme_bw() 
# V2.plot

```

```{r}
# Create Palmprint - palmprint alignment-distance network
palm.graph <- graph.palm( virome.df )
  vir2palm <- match(V(palm.graph)$name, V(vir.graph2)$name) # look-up index
  
  V(palm.graph)$pr    <- V(vir.graph2)$pr[ vir2palm ]
  V(palm.graph)$vrich <- V(vir.graph2)$vrich[ vir2palm ]
  V(palm.graph)$vrank <- V(vir.graph2)$vrank[ vir2palm ]
  V(palm.graph)$tax_species <- V(vir.graph2)$tax_species[ vir2palm ]

  
#palm.graph2 <- delete_vertices( palm.graph, V(palm.graph)$name[ (degree(palm.graph) < 2) ])
createNetworkFromIgraph(palm.graph,"palmGraph")         
```

```{r}
  
# Plot the Degree of interconnectivity for Palmprint network
palm.degree <- data.frame( set = "observed", setn = '0',
                           degree = as.numeric(degree(palm.graph)))

# Generate random-sampled Control Sets of Palmprints
# (sample N palmprints from all palmprints)
n.controlsets = 10

for (i in 1:n.controlsets){
  ctrl.g <- graph.palmControl(virome.df)
  palm.degree <- rbind( palm.degree,
                        data.frame( set = "expected",
                                    setn = as.character(i),
                                    degree = as.numeric(degree(ctrl.g))) )
}


# Rank Observed and Expected (sampled) Palmprint-sets by the max-degrees of a node
observed.rank <- palm.degree$rank[ palm.degree$set == 'observed'][1] # store observed rank value
print("The Observed Network contains:  ")

palm.degree.max <- aggregate(palm.degree, by = list(palm.degree$setn), max)
  palm.degree.max <- palm.degree.max[ , c('setn', 'degree')]
  palm.degree.max <- palm.degree.max[ order(palm.degree.max$degree, decreasing = TRUE), ]
  palm.degree.max$rank <- 1:length(palm.degree.max$setn)
  # assign ranks
  palm.degree$rank <- palm.degree.max$rank[ match(palm.degree$setn, palm.degree.max$setn) ] 
  
  
```

```{r}

rd.plot <- ggplot( palm.degree, aes(degree,
                                    alpha = set,
                                    fill  = setn)) +
  geom_histogram(binwidth = 1,  position = 'identity', colour = NA) +
  theme_bw() + xlab('sOTU node degree') +
  theme(legend.position = "none") +
  scale_alpha_manual(  values = c(0.05, 0.5) ) +
  scale_fill_manual( values = c( 'orange2',  rep("black", length(unique(palm.degree$setn)) -1 ))) +
  facet_wrap(~set, ncol = 1)
rd.plot

rd.plot <- ggplot( palm.degree, aes(rank, degree,
                                    color = set,
                                    fill = setn,
                                    alpha = set)) +
  geom_jitter(height = 0.5, width = 0.2) +
  geom_boxplot(color = "black", outliers =  FALSE) +
  scale_alpha_manual(values = c(0.4, 0.4)) +
  scale_color_manual(values = c('gray50', 'orange2')) +
  scale_fill_manual( values = c( 'orange2',  rep("gray70", length(unique(palm.degree$setn)) -1 ))) +
  theme_bw() + xlab('Rank set') +
  theme(legend.position = "none")
rd.plot

palm.degree.count <- melt(table( palm.degree[ , c("rank", "degree")]))
  palm.degree.count$set <- "expected"
  palm.degree.count$set[ palm.degree.count$rank == observed.rank] <- "observed"

rd.plot <- ggplot( palm.degree.count, aes(x = factor(degree),
                                          y = value,
                                          color = set)) +
  geom_boxplot(outliers = FALSE) +
  geom_jitter(height = 0, alpha = 0.05) +
  scale_alpha_manual(values = c(0.4, 0.4)) +
  scale_color_manual(values = c('gray50', 'orange2')) +
  theme_bw() + xlab('Node Degree') +
  theme(legend.position = "none") + coord_flip()
rd.plot

# Control RdRp Network
# createNetworkFromIgraph(ctrl.g,"rdrp")            

```

```{r}
## Not working with >20 bioProjects
# # Create a data.frame to plot a heatmap of Virus Family Expression
# # where the heatmap value is the Number of BioProject runs positive for that Virus
# 
# # Create virome family heatmap (top 20 Taxa)
# virFam.df <- virome.df[,  c('tax_family', 'bio_project')]
# virFam.df$tax_family <- makeTop10(virFam.df$tax_family, top.n = 20)
# virFam.df <- table(virFam.df)
#   bpTop   <- colnames(virFam.df)
#   bpTop.n <- bp.total.n$n[ match(bpTop, bp.total.n$bioproject) ]
# 
# # Divide each matrix-row by vector
# # Form: t( t(M) / vec )
# virFam.df2 <- t( t(virFam.df) / bpTop.n)
# virFam.df2 <- round( 100 * virFam.df2, 2 )
# virFam.df2 <- as.matrix(virFam.df2)
#   
# # Virome Family count bar plot
# virFam.bp <- gplots::heatmap.2(virFam.df2, trace = "none", 
#                                breaks = c(0,1, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100),
#                                density.info="none",
#                                col = c("black", viridis(10, option = "A")),
#                                key.title = "",
#                                key.xlab = "Percent BioProject in Target",
#                                margins = c(10,10), sepcolor=NULL)
```



```{r, echo = FALSE}
# Create virome organism summary data.frame
vorgx.df2 <- virome.df2[ virome.df2$node_qc, ] %>%
              count(scientific_name, tax_family, sort = TRUE)
  vorgx.df2$scientific_name <- factor(vorgx.df2$scientific_name,
                                     levels = rev( levels(virome.df2$scientific_name)))
  vorgx.df2$tax_family      <- factor(vorgx.df2$tax_family,
                                     levels = levels(virome.df2$tax_family ))
  vorgx.df2 <- vorgx.df2[ !is.na(vorgx.df2$scientific_name), ]

# Virome Tax Distribution
virome2.org <- ggplot(vorgx.df2, aes( x = tax_family, n,
                                    fill = tax_family)) +
  geom_bar( stat = 'identity') +
  scale_y_log10() + coord_polar("x", start = 0) +
  theme_bw() + 
  theme(aspect.ratio = 1, legend.position = "none") + 
  facet_wrap(~scientific_name, ncol = 4)

```


### Label Summary {#meta-data-module}

```{r,  fig.height = 8, fig.width = 10, retina = 3}
# Per "Species" family-levle virome summary
virome2.org
```

#### Per Virus-sOTU Summary - Expression and Frequency

```{r, echo = FALSE}
# Create plot-color scheme

# phylum, family, genus, species
  ranklvl <-  c("phylum",  "family",  "genus",   "species")

  rankcols <- c("phylum" = "#9f62a1",
                "family" = "#00cc07",
                "genus"  = "#ff9607",
                "species"= "#ff2a24")

  # Add color highlight to hits within the same phylum
  virx.df$gb_match <- ranklvl[1]
  virx.df$gb_match[ which( virx.df$gb_pid >= 45 ) ] <- ranklvl[2]
  virx.df$gb_match[ which( virx.df$gb_pid >= 70 ) ] <- ranklvl[3]
  virx.df$gb_match[ which( virx.df$gb_pid >= 90 ) ] <- ranklvl[4]

  virx.df$gb_match <- factor(virx.df$gb_match,
                         levels = ranklvl)


# Virus Abundance and Expression
virus.exp <- ggplot() + 
  geom_point( data = virx.df, aes(n, mean_coverage,
                                  pp    = sotu,
                                  gbacc = gb_acc,
                                  gbid  = gb_pid,
                                  tax   = tax_species,
                                  fam   = tax_family,
                                  color = gb_match ),
              show.legend = FALSE,
              alpha = 0.8, size = 3) +
  geom_rug(outside = FALSE) +
  theme_bw() +
  scale_color_manual(values = rankcols, drop = FALSE) +
  scale_x_log10() + scale_y_log10() +
  xlab("sOTU frequency in SRA Runs") +
  ylab("sOTU mean contig coverage")

library("viridis")
virx.df$plot_name <- makeTop10(virx.df$tax_family)

# Virus Abundance and Expression
virus.exp2 <- ggplot() + 
  geom_point( data = virx.df, aes(x = n, y = gb_pid,
                                  size = log(mean_coverage + 1),
                                  pp    = sotu,
                                  covrg = mean_coverage,
                                  gbacc = gb_acc,
                                  gbid  = gb_pid,
                                  tax   = tax_species,
                                  fam   = tax_family,
                                  color = log(mean_coverage + 1)),
              show.legend = FALSE,
              alpha = 0.5) +
  geom_hline( yintercept = 90, color = "gray70", linetype = "dashed" ) + 
  #geom_rug(outside = FALSE) +
  theme_bw() +
  scale_color_viridis(option = "plasma") +
  scale_x_log10() + scale_y_log10() + scale_size_identity() +
  xlab("sOTU frequency in SRA Runs") +
  ylab("sOTU mean contig coverage") +
  facet_wrap(~plot_name, ncol = 4)

#virus.exp2

```

```{r}
# Virus Expression and SRA-Frequency
plotly::hide_legend( virus.exp2 )
```

```{r, echo = FALSE}
# Create univariable histograms to see underlying distrubition
virx.df$tax_family2 <- makeTop10(virx.df$tax_family)

virus.hist.cov <- ggplot( virx.df, aes(mean_coverage,
                                       fill = tax_family2) ) +
  geom_histogram() +
  scale_x_log10() +
  theme_bw()

virus.hist.n   <- ggplot( virx.df, aes(n,
                                       fill = tax_family2) ) +
  geom_histogram() +
  scale_x_log10() +
  theme_bw()

virus.hist.gbid <- ggplot( virx.df, aes(gb_pid,
                                       fill = tax_family2) ) +
  geom_histogram() +
  scale_x_log10() +
  theme_bw()

```

```{r}
# Virus (sOTU) summary statistics histogram
# Frequency in SRA
virus.hist.n
# Mean coverage of contigs
virus.hist.cov
# Alignment identity to nearest GenBank record
virus.hist.gbid

```

#### Virome Data Tables

- `full_virome.csv`

```{r, echo = FALSE}
### Data-Table: `virome.csv`
# Data table for Full Virome table

blast.col <- linkBLAST(header = paste0(virome.df$run,"_",
                                       virome.df$palm_id,"_",
                                       virome.df$nickname),
                       aa.seq = virome.df$node_seq )

sra.col       <- linkDB(virome.df$run)
biosample.col <- linkDB(virome.df$bio_sample, DB = "biosample")

cbind( as.character(virome.df$scientific_name),
       sra.col, biosample.col,
       as.character(virome.df$bio_project),
       as.character(virome.df$sotu),
       as.character(virome.df$nickname),
       virome.df$node_coverage,
       as.character(virome.df$gb_acc),
       virome.df$gb_pid,
       as.character(virome.df$tax_species),
       as.character(virome.df$tax_family),
       blast.col) %>%
  DT::datatable(
    colnames = c( "scientific_name", "sra_run", "biosample_id", "bioproject",
                 "sOTU", 'nickname', 'coverage', "gb_accession", "gb_id%", "tax_species", "tax_family", "BLAST"),
    rownames = FALSE, filter = "top", escape = F,
    options = list(pageLength = 20, scrollX = T,
                   order = list(list(6, 'desc')))
  )

```

- `summary_virome.csv`

```{r, echo = FALSE}
### Data-Table: `summary_virome.csv`
# Data table for Summary Virome table

cbind( as.character(virx.df$sotu),
       as.character(virx.df$nickname),
       virx.df$n,
       virx.df$mean_coverage,
       as.character(virx.df$gb_acc),
       as.character(virx.df$gb_pid),
       as.character(virx.df$tax_species),
       as.character(virx.df$tax_family)
       ) %>%
  DT::datatable(
    colnames = c("sOTU", "nickname",
                 "n_SRA_runs", "mean_coverage", "gb_accession", "gb_id%", "tax_species", "tax_family"),
    rownames = FALSE, filter = "top", escape = F,
    options = list(ordering = TRUE, order = list(list(2, 'desc')),
                   pageLength = 20, scrollX = T)
  )

```


<!-- # Bespoke Analysis ------------------------------------------------------------- -->
<!-- ```{r} -->
<!-- virome.stat <- virome.stat[ order(virome.stat$total_count), ] -->
<!-- virome.stat$logCount <- log10(virome.stat$total_count) -->
<!-- # STAT Read Plot -->
<!-- statReads <- ggplot( virome.stat, aes(logCount) ) + -->
<!--   geom_histogram(bins = 50) + xlab('log10 (STAT Read Count)') + ylab('N SRA Runs') -->
<!-- statReads -->
<!-- ``` -->
<!-- ```{r} -->
<!-- # Candida-Astrovirus analysis -->
<!-- meta.df <- read.csv('ov/candida/PRJNA827179.normalized.csv', header = T) -->
<!-- meta.df$Condition = factor(meta.df$Condition) -->
<!-- meta.df$Condition  <- factor( meta.df$Condition, -->
<!--                               levels =  c( "WT", "UME6Off", "UME6On", -->
<!--                                            "Ume6OfftoOn_0hr",  "Ume6OfftoOn_3hr", "Ume6OfftoOn_6hr", "Ume6OfftoOn_12hr", -->
<!--                                            "Ume6OntoOff_0hr", "Ume6OntoOff_3hr", "Ume6OntoOff_6hr",   "Ume6OntoOff_12hr" )) -->
<!-- astro.df <- virome.df[ (virome.df$tax_family  == 'Astroviridae'), ] -->
<!-- astro.df <- astro.df[ !is.na(astro.df$run), ] -->
<!-- meta.df <- merge(meta.df, astro.df, by.x = 'biosample_id', by.y = 'bio_sample', all.x = TRUE) -->
<!-- #astro.df <- astro.df[ !is.na(astro.df$run), ] -->
<!-- # astroPlot <- ggplot( astro.df[ astro.df$bio_project == 'PRJNA827179', ], -->
<!-- #                      aes(total_count, node_coverage) ) + -->
<!-- #   geom_point() + theme_bw() -->
<!-- # astroPlot -->
<!-- astroPlot <- ggplot( meta.df, aes(c_RPM, v_RPM) ) + -->
<!--   geom_point() + theme_bw() + -->
<!--    xlab('Candida_ReadsPerMillion') + -->
<!--    ylab('Astrovirus_ReadsPerMillion') -->
<!-- astroPlot -->
<!-- cor(meta.df$v_RPM, meta.df$c_RPM) -->
<!-- astroBar <- ggplot( meta.df, -->
<!--                     aes(Condition, v_RPM, fill = Condition) ) + -->
<!--   geom_boxplot() + geom_jitter() + theme_bw() + -->
<!--    ylab('Astrovirus_ReadsPerMillion') -->
<!-- astroBar -->
<!-- astroBar2 <- ggplot( meta.df, -->
<!--                     aes(Condition, c_RPM, fill = Condition) ) + -->
<!--   geom_boxplot() + geom_jitter() + theme_bw() + -->
<!--   ylab('Candida_ReadsPerMillion') -->
<!-- astroBar2 -->
<!-- ``` -->
<!-- ```{r} -->
<!-- rpmPlot <- ggplot( virome.df, aes(node_coverage, total_count, -->
<!--                                   color = tax_family) ) + -->
<!--   geom_point() + scale_x_log10() + scale_y_log10() + -->
<!--   facet_wrap(~tax_family) + theme_bw() + theme(legend.position = "none") -->
<!-- rpmPlot -->
<!-- ``` -->



